core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeCompactionSuite.scala

```diff
@@ -184,6 +219,53 @@ trait OptimizeCompactionSuiteBase extends QueryTest
     }
   }
 
+
+  test("optimize command via DeltaTable: on partitioned table - selected partitions") {
+    withTempDir { tempDir =>
+      val path = new File(tempDir, "testTable").getCanonicalPath
+      val partitionColumns = Some(Seq("id"))
+      appendToDeltaTable(
+        Seq(1, 2, 3).toDF("value").withColumn("id", 'value % 2),
+        path,
+        partitionColumns)
+
+      appendToDeltaTable(
+        Seq(4, 5, 6).toDF("value").withColumn("id", 'value % 2),
+        path,
+        partitionColumns)
+
+      val deltaLogBefore = DeltaLog.forTable(spark, path)
+      val txnBefore = deltaLogBefore.startTransaction();
+      val fileListBefore = txnBefore.filterFiles()
+
+      assert(fileListBefore.length >= 3)
+      assert(fileListBefore.count(_.partitionValues === Map("id" -> "0")) > 1)
+
+      val versionBefore = deltaLogBefore.snapshot.version
+      io.delta.tables.DeltaTable.forPath(spark, path).optimize("id = 0")
```

Comment:
One suggestion to minimize the test code duplication.

Add an abstract method `executeOptimize(path, condition)` to `OptimizeCompactionSuiteBase` and modify all tests to call this method rather than invoking the SQL or Scala API directly.

Rename `OptimizeCompactionSuite` to `OptimizeCompactionSQLSuite` and implement `executeOptimize(path, condition)` that issues a Optimize SQL command.

Add `OptimizeCompactionScalaSuite` extending `OptimizeCompactionSuite` and implement `executeOptimize(path, condition)` that issues a Optimize Scala API call.

Reply:
I'll try that out

Reply:
It can be done similar to `executeGenerate` [here](https://github.com/delta-io/delta/blob/master/core/src/main/scala/io/delta/tables/execution/DeltaTableOperations.scala#L57). Instead of using the `OptimizeExecutor`, use the `DeltaCommand` implementation `OptimizeTableCommand`. This also removes the changes where the partition filter has is moved from `OptimizeTableCommand` to `OptimizeExecutor`.

For vacuum, please create an issue.

Reply:
Got that working, made two funcs to handle path or table